ui/components/app/simulation-details/useBalanceChanges.ts

```diff
@@ -0,0 +1,175 @@
+  decimals: number,
+): Amount => {
+  const numeric = Numeric.from(quantity, 16)
+    .times(isNegative ? -1 : 1, 10)
+    .toBase(10)
+    .shiftedBy(decimals);
+  return { isNegative, quantity, decimals, numeric };
+};
+
+// Fetches token details for all the token addresses in the SimulationTokenBalanceChanges
+const fetchErc20Decimals = async (
+  addresses: Hex[],
+): Promise<Record<Hex, number>> => {
+  const uniqueAddresses = [...new Set(addresses)];
+  const tokenInfos = await Promise.all(
+    uniqueAddresses.map((address) => getTokenStandardAndDetails(address)),
+  );
+
+  return tokenInfos.reduce(
+    (result, { decimals }, index) => ({
+      ...result,
+      [uniqueAddresses[index]]: decimals
+        ? parseInt(decimals, 10)
+        : ERC20_DEFAULT_DECIMALS,
+    }),
+    {},
+  );
+};
+
+// Compiles the balance change for the native asset
+const getNativeBalanceChange = (
+  nativeBalanceChange: SimulationBalanceChange | undefined,
+  nativeFiatRate: number,
+): BalanceChange | undefined => {
+  if (!nativeBalanceChange) {
+    return undefined;
+  }
+  const asset = NATIVE_ASSET;
+  const amount = getAssetAmount(nativeBalanceChange, NATIVE_DECIMALS);
+  const fiatAmount = amount.numeric
+    .applyConversionRate(nativeFiatRate)
+    .toNumber();
+  return { asset, amount, fiatAmount };
+};
+
+// Compiles the balance changes for token assets
+const getTokenBalanceChanges = (
+  tokenBalanceChanges: SimulationTokenBalanceChange[],
+  erc20Decimals: Record<Hex, number>,
+  tokenFiatRates: Record<Hex, number>,
+): BalanceChange[] => {
+  return tokenBalanceChanges.map((tokenBc) => {
+    const asset: TokenAssetIdentifier = {
+      standard: convertStandard(tokenBc.standard),
+      address: tokenBc.address.toLowerCase() as Hex,
+      tokenId: tokenBc.id,
+    };
+
+    const decimals =
+      asset.standard === TokenStandard.ERC20 ? erc20Decimals[asset.address] : 0;
+    const amount = getAssetAmount(tokenBc, decimals);
+
+    const fiatRate = tokenFiatRates[tokenBc.address];
+    const fiatAmount = fiatRate
+      ? amount.numeric.applyConversionRate(fiatRate).toNumber()
+      : FIAT_UNAVAILABLE;
+
+    return { asset, amount, fiatAmount };
+  });
+};
+
+// Get the exchange rates for converting tokens to the user's fiat currency.
+const getTokenToFiatConversionRates = createSelector(
+  getTokenExchangeRates,
+  getConfirmationExchangeRates,
+  (contractExchangeRates, confirmationExchangeRates) => {
+    const mergedRates = {
+      ...contractExchangeRates,
+      ...confirmationExchangeRates,
+    } as Record<string, number>;
+    return Object.entries(mergedRates).reduce((acc, [key, value]) => {
+      acc[key.toLowerCase()] = value;
+      return acc;
+    }, {} as Record<string, number>);
+  },
+);
+
+// Compiles a list of balance changes from simulation data
+export const useBalanceChanges = (
+  simulationData?: SimulationData,
+): { pending: boolean; value: BalanceChange[] } => {
+  if (!simulationData) {
```

Comment:
I thought React stopped you returning before any hooks as it wanted to run all hooks on every render? Or is that just conditionals?

Reply:
I'm not sure why the code has been working. I've modified it to run the hooks first before the early return.